### PR TITLE
Revive the script: use newer API and getHistory method

### LINKED ISSUE #7: VK removes out params from messages.get
VK say they will remove `out` param from `messages.get` requests after a month. I asked they how to get incoming messages in the case.

### PR DESCRIPTION
**Pull request by karlicoss**
The messages.get api method is pretty much dead... so I switched the script to newer API and made sure everything still works with the messages retrieved from old API. I diffed the json files in my storage, and it looks ok.

Another fix is the `sort` method to sort by id instead of datetime, since sometimes timestamps are equal, so they are not reliable for sorting.

Should fix https://github.com/Totktonada/vk_messages_backup/issues/7

### CODE REVIEW

**Comment by Totktonada** [comment]
@karlicoss Thank you! I'll review on the week.

File: vk_messages_backup.py [hunk 1]
```diff
@@ -148,17 +148,32 @@ def do_request(self, method, params):
...
 def get_msg_body(m):
     body = m.get('body', None)
     if body is not None:
         return body
     return m.get('text', None)
 
+
 def get_chat_title(m):
-    return m.get('title', 'NO TITLE') # TODO
+    # TODO in new api, title is not sent in each chat message
```
Comment (Totktonada) [inline vk_messages_backup.py:173]: Old does not do it too, but AFAIR there were messages with titles.

File: vk_messages_backup.py [hunk 2]
```diff
@@ -204,25 +220,35 @@ def format_forward(msg):
...
         def format_action(msg):
             if 'action' in msg:
+                action = msg['action']
+                atype: str
+                act_username = None
+                if isinstance(action, dict): # new api
+                    atype = action['type']
+                    if 'member_id' in action:
+                        mid = action['member_id']
+                        act_username = format_username_by_id(mid)
```
Comment (Totktonada) [inline vk_messages_backup.py:230]: `action` can contain `email` field in the new API, it is the same as `action_email` in the old one.

File: vk_messages_backup.py [hunk 3]
```diff
@@ -204,25 +220,35 @@ def format_forward(msg):
...
-                    else:
-                        act_username = msg['action_email']
-                if msg['action'] == 'chat_photo_update':
+                if atype == 'chat_photo_update':
                     action = 'chat photo updated'
-                elif msg['action'] == 'chat_photo_remove':
+                elif atype == 'chat_photo_remove':
                     action = 'chat photo removed'
-                elif msg['action'] == 'chat_create':
+                elif atype == 'chat_create':
```
Comment (Totktonada) [inline vk_messages_backup.py:245]: The new API has no `action_text` field, it is `text` field in `action` object.

File: vk_messages_backup.py [hunk 4]
```diff
@@ -433,11 +462,17 @@ def load(self):
...
+        did = peer_id_to_dialog_id(peer_id)
+        dialog = self.dialogs.get(did, None)
+        if dialog is None:
+            return None
+
+        messages = dialog.messages
+        if len(messages) == 0:
+            return None
+
+        return max(m.id() for m in messages)
```
Comment (Totktonada) [inline vk_messages_backup.py:468]: I guess we can just pick a last message.

File: vk_messages_backup.py [hunk 5]
```diff
@@ -564,39 +554,29 @@ def users_dict(self, my_id):
...
-        for msg in messages:
-            if msg.id() not in ids:
-                res_messages.append(msg)
-                ids.add(msg.id())
-        params['offset'] += msg_per_request
-    return res_messages
+    done = False
+    while not done:
+        if len(res) > 0:
+            params['start_message_id'] = res[-1]['id'] - 1
```
Comment (Totktonada) [inline vk_messages_backup.py:569]: Why -1? BTW, I guess it is easier to iterate with `offset`.

**Review by Totktonada — COMMENTED** [review]
Sorry for the late response.

I'm look at the PR, started to rewrite some parts, but found that now I has no access to the VK API, because of https://vk.com/dev/messages_api . Acquired the access and waiting for it.

Meanwhile I have pointed several things I would change if everything will be ok with accesses and I'll have free time.

I would not bother about old format compatibility: it looks error-prone. Also json dumps in the old format are likely broken, because they lack of sent messages, so we cannot lean on a last message id from an old json dumps. I would add a file within `storage` directory with a version of these files and skip them if this file is not exists.

I would fix all flake8 reported problems.

It seems we should also change the VK API version in the link in README.

@karlicoss Don't sure how to better proceed further. Let me know if you want to continue work on this PR.

**Comment by Totktonada** [comment]
My request for the access was declined. VK support says that they maybe will add read-only access for standalone apps in a future, but explicitly state that cannot provide any dates.

For now this project should be considered as abandoned.

**Comment by karlicoss** [comment]
Yup, I ran into the same issue at some point, my fixed script worked for couple of weeks till they revoked the endpoint ;(

**Comment by Totktonada** [comment]
I've added a fair status report: https://github.com/Totktonada/vk_messages_backup/commit/a0241bb2a6df600a9a52347f210ee5ac43cb5d7b. I have no plans to fix/maintain the script.

The license (public domain) allows to copy and distribute it in any way (but unlikely it would be useful for anyone).

So I'm going to close issues/PRs. I'll accept issues/PRs, which suggest better alternatives.

**Pull request closed without merging**